### ISSUE TITLE
Override module resource from outside module

### ISSUE THREAD

**Issue by nitrocode** [hashicorp/terraform#27525]
Instead of having to wait for a module owner to update a resource in their tf module and instead of having to fork the module itself to make those changes, it would be nice to override the module resource from outside the module so you have full access to all of the resource-to-override's arguments and still use most of the feature set that comes from using the module itself.

**Comment by elvis2** [comment]
I second this request. What is the point of using 3rd provider modules if you can not override them? As the world begins to modernize their IaC stacks, using 3rd party modules to create a more common stack templated approach, this override feature becomes critical.

**Comment by asaranglr** [comment]
Any updates on this? Seems like a neat feature to have.

**Comment by danielebra** [comment]
Any updates on this issue? This feature would be beneficial for using established modules that are lacking an additional modification.

**Comment by pkrolikowski** [comment]
It's tempting to override resource attributes from outside the module, but it will brake "encapsulate configuration" promise.
Very often modules are created to force user to create infrastructure by the rules defined by the module. Enabling overriding resource definition from outside is a bad thing in this case.

For now you can use [override logic](https://www.terraform.io/language/files/override) and yes, you have to fork but maintaining is much easier.

**Comment by waltervargas** [comment]
Is there any updates on this? can you provide a mechanism similar to the `outputs`, maybe an `export` keyword, so you allow the caller to override resources in the module.

**Comment by Z4ck404** [comment]
Are there any updates about this ?

**Comment by simoncpu** [comment]
Dear Santa Claus,

I wish for the ability to override module resources from outside the modules too.

Thanks, Santa!

**Comment by nitrocode** [comment]
There is one way to do it now 

1. `terraform init`
1. Use `patch` to apply a local patch to `.terraform` dir
1. `terraform plan` or `terraform apply`

**Comment by nitrocode** [comment]
Overriding a module with a local patch is similar to how you can use kustonize to use an upstream helm chart and apply a patch.

We're basically asking for a kustomize for terraform but built-in so we don't have to build the feature into yet another terraform wrapper like terragrunt, atmos, terramate, etc.

**Comment by simoncpu** [comment]
This sounds interesting. I wish this would work on Terraform Cloud though...

**Comment by simoncpu** [comment]
My use case is that I reorganized my code into two modules. Both modules
call `cloudflare_zone` on the same zone. I'm looking for a way to let the
main code override the resources on both modules so that Terraform would
treat it as a single resource.

**Comment by benamazing** [comment]
Yes, that would be great if terraform provides a feature like kustomize patch.

**Comment by kderck** [comment]
This would still be great.

**Comment by Saurabhbatham** [comment]
This feature is required, it will help to customize the module resources on which you don't have control.

**Comment by brannondorsey** [comment]
There are many times where I've wished to reach for a feature like this. I recognize it can break some of the beauty of a well-crafted module encapsulation/abstraction, but it's not uncommon to find yourself in a situation where it just makes practical sense to be able to craft an override for a module you don't control. This is especially the case in scenarios where vendoring a fork of someone else's module actually introduces more complexity than a well crafted and commented override would.

**Comment by sanjeevpandey19** [comment]
+1 for this feature

**Comment by crw** [comment]
Thanks for your interest in this issue! This is just a reminder to please avoid "+1" comments, and to use the upvote mechanism (click or add the 👍 emoji to the original post) to indicate your support for this issue. This helps avoid notification spam for issues with high numbers of participants while enabling the maintainers to prioritize issues. Thanks again for the feedback!

**Comment by liamhelmer** [comment]
Agreed, there's a lot of use cases that this would improve! Please add this in.

**Comment by dror-g** [comment]
Yea, don't hold your breath.. seems other similar issues close as `not planned / don't care` .

> Since this feature would require significant work and be potentially dangerous to users, we have decided to not implement this feature at this time and will be closing the issue.

_Originally posted by @ghost in https://github.com/opentofu/opentofu/issues/808#issuecomment-1968337057_

**Comment by mihir83in** [comment]
Could it be possible to add restricted override behaviour without compromising the integrity of the terraform module?

If we could define attributes of resource, or resource overridable then it can be reasoned about in plan or when overriding. And it can be expected that any code dependent on such overridable or variant resource in a module would make the module integrity lower numbered.

It can be a great useful feature to allow composition.
